src/Compilers/Core/Portable/Text/ChangedText.cs

```diff
@@ -316,29 +333,33 @@ private static ImmutableArray<TextChangeRange> Merge(ImmutableArray<TextChangeRa
                             AddRange(list, oldChange);
                             oldDelta = oldDelta - oldChange.Span.Length + oldChange.NewLength;
                             oldIndex++;
+                            needNextOldChange = true;
                             goto nextOldChange;
                         }
                         else if (newChange.Span.Length <= oldChange.NewLength)
                         {
                             // new change deletes fewer characters than old change inserted
-                            // add new change insertion, then the remaining trailing characters of the old change insertion
-                            AddRange(list, new TextChangeRange(oldChange.Span, oldChange.NewLength + newChange.NewLength - newChange.Span.Length));
-                            oldDelta = oldDelta - oldChange.Span.Length + oldChange.NewLength - newChange.Span.Length;
-                            oldIndex++;
+                            // apply the new change, and adjust the old change
+                            var adjustedNewChange = new TextChangeRange(new TextSpan(newChange.Span.Start - oldDelta, newChange.Span.Length), newChange.NewLength);
+                            AddRange(list, adjustedNewChange);
+                            oldChange = new TextChangeRange(new TextSpan(oldChange.Span.Start + adjustedNewChange.Span.Length, oldChange.Span.Length - adjustedNewChange.Span.Length), oldChange.NewLength - newChange.Span.Length);
                             newIndex++;
+                            needNextNewChange = true;
                             goto nextNewChange;
```

Comment:
📝 Most of the complexity of this change involves the new ability to `goto nextNewChange` without overwriting the value of `oldChange`, since the value might be a subset of the change in `oldChanges[oldIndex]`.